core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -4560,6 +4593,39 @@ object Stream extends StreamLowPriority {
     def toVector: G[Vector[O]] = to(Vector)
   }
 
+/**
+  * An abstraction for writing `Pull` computations that can timeout
+  * while reading from a `Stream`.
+  *
+  * A `TimedPull` is not created or intepreted directly, but by calling [[timed]].
+  * {{{
+  * yourStream.pull.timed(tp => ...).stream
+  * }}}
+  *
+  * The argument to `timed` is a `TimedPull[F, O] => Pull[F, O2, R]`
+  * function which describes the pulling logic, and is often recursive, with shape:
+  *
+  * {{{
+  * def go(tp: TimedPull[F, A]): Pull[F, B, Unit] =
+  *   tp.uncons.flatMap {
+  *     case Some((Right(chunk), next)) => doSomething >> go(next)
+  *     case Some((Left(_), next)) => doSomethingElse >> go(next)
+  *     case None => Pull.done
+  *   }
+  * }}}
+  *
+  * Where `doSomething` and `doSomethingElse` are `Pull` computations
+  * such as `Pull.output`, in addition to `TimedPull.timeout`.
+  *
+  * See below for detailed descriptions of `timeout` and `uncons`, and
+  * look at the [[ToPull.timed]] scaladoc for an example of usage.
+  */
+  trait TimedPull[F[_], O] {
```

Comment:
Thoughts on defining this as `Pull.Timed` instead of `Stream.TimedPull`?

Reply:
oh interesting. It is closer to `ToPull` in spirit, but I do like `Pull.Timed` and it's probably more discoverable, so if you're definitely :+1: on `Pull.Timed`, I'll move it there :)